### PR TITLE
nuttx/timer: add arch_alarm tick interface

### PR DESCRIPTION
**Pull request by zyfeier**
## Summary

1. Use tick count for sched timer expiration;
2. Adjust arch_alarm to support tick;

## Impact

NA

## Testing

sabre-6quad:nsh

### CODE REVIEW

**Comment by acassis** [comment]
Hi @zyfeier please include more information about this PR, what was the main motivation and what are the benefits of using this new interface? Did it make the system more accurate for timer measuring?

**Comment by zyfeier** [comment]
> Hi @zyfeier please include more information about this PR, what was the main motivation and what are the benefits of using this new interface? Did it make the system more accurate for timer measuring?

Hi,  acassis. This PR is not ready for review, it's a temporary change. I'll add more information if it was ready.

File: arch/arm/src/stm32/stm32_tickless.c [hunk 1]
```diff
@@ -698,9 +698,9 @@ int up_timer_gettime(struct timespec *ts)
  *
  ****************************************************************************/
 
-int up_timer_getcounter(uint64_t *cycles)
+int up_timer_gettick(FAR clock_t *ticks)
```
Comment (pkarashchenko) [inline arch/arm/src/stm32/stm32_tickless.c:701]:
```suggestion
int up_timer_gettick(clock_t *ticks)
```

File: arch/arm/src/stm32/stm32_tickless.c [hunk 2]
```diff
@@ -718,7 +718,7 @@ int up_timer_getcounter(uint64_t *cycles)
  *
  ****************************************************************************/
 
-void up_timer_getmask(uint64_t *mask)
+void up_timer_getmask(FAR clock_t *mask)
```
Comment (pkarashchenko) [inline arch/arm/src/stm32/stm32_tickless.c:721]:
```suggestion
void up_timer_getmask(clock_t *mask)
```

File: arch/arm/src/stm32f7/stm32_tickless.c [hunk 3]
```diff
@@ -756,9 +756,9 @@ int up_timer_gettime(struct timespec *ts)
  *
  ****************************************************************************/
 
-int up_timer_getcounter(uint64_t *cycles)
+int up_timer_gettick(FAR clock_t *ticks)
```
Comment (pkarashchenko) [inline arch/arm/src/stm32f7/stm32_tickless.c:759]:
```suggestion
int up_timer_gettick(clock_t *ticks)
```

File: arch/arm/src/stm32f7/stm32_tickless.c [hunk 4]
```diff
@@ -776,7 +776,7 @@ int up_timer_getcounter(uint64_t *cycles)
  *
  ****************************************************************************/
 
-void up_timer_getmask(uint64_t *mask)
+void up_timer_getmask(FAR clock_t *mask)
```
Comment (pkarashchenko) [inline arch/arm/src/stm32f7/stm32_tickless.c:779]:
```suggestion
void up_timer_getmask(clock_t *mask)
```

File: arch/arm/src/stm32h7/stm32_tickless.c [hunk 5]
```diff
@@ -729,9 +729,9 @@ int up_timer_gettime(struct timespec *ts)
  *
  ****************************************************************************/
 
-int up_timer_getcounter(uint64_t *cycles)
+int up_timer_gettick(FAR clock_t *ticks)
```
Comment (pkarashchenko) [inline arch/arm/src/stm32h7/stm32_tickless.c:732]:
```suggestion
int up_timer_gettick(clock_t *ticks)
```

File: arch/arm/src/stm32h7/stm32_tickless.c [hunk 6]
```diff
@@ -749,7 +749,7 @@ int up_timer_getcounter(uint64_t *cycles)
  *
  ****************************************************************************/
 
-void up_timer_getmask(uint64_t *mask)
+void up_timer_getmask(FAR clock_t *mask)
```
Comment (pkarashchenko) [inline arch/arm/src/stm32h7/stm32_tickless.c:752]:
```suggestion
void up_timer_getmask(clock_t *mask)
```

File: arch/arm/src/stm32wb/stm32wb_tickless.c [hunk 7]
```diff
@@ -572,9 +572,9 @@ int up_timer_gettime(struct timespec *ts)
  *
  ****************************************************************************/
 
-int up_timer_getcounter(uint64_t *cycles)
+int up_timer_gettick(FAR clock_t *ticks)
```
Comment (pkarashchenko) [inline arch/arm/src/stm32wb/stm32wb_tickless.c:575]:
```suggestion
int up_timer_gettick(clock_t *ticks)
```

File: arch/arm/src/stm32wb/stm32wb_tickless.c [hunk 8]
```diff
@@ -592,7 +592,7 @@ int up_timer_getcounter(uint64_t *cycles)
  *
  ****************************************************************************/
 
-void up_timer_getmask(uint64_t *mask)
+void up_timer_getmask(FAR clock_t *mask)
```
Comment (pkarashchenko) [inline arch/arm/src/stm32wb/stm32wb_tickless.c:595]:
```suggestion
void up_timer_getmask(clock_t *mask)
```

File: drivers/timers/arch_timer.c [hunk 9]
```diff
@@ -315,6 +302,21 @@ int weak_function up_timer_gettime(FAR struct timespec *ts)
...
 
+#if defined(CONFIG_SCHED_TICKLESS_TICK_ARGUMENT) || defined(CONFIG_CLOCK_TIMEKEEPING)
+int weak_function up_timer_gettick(FAR clock_t *ticks)
+{
+  int ret = -EAGAIN;
+
+  if (g_timer.lower != NULL)
+    {
+      *ticks = current_usec() / USEC_PER_TICK;
+      ret = 0;
```
Comment (pkarashchenko) [inline drivers/timers/arch_timer.c:313]:
```suggestion
      ret = OK;
```

File: include/nuttx/clock.h [hunk 10]
```diff
@@ -251,6 +251,19 @@ EXTERN volatile clock_t g_system_ticks;
  * Public Function Prototypes
  ****************************************************************************/
 
+#define timespec_from_tick(ts, tick) \
+  do \
+    { \
+      clock_t _tick = tick; \
```
Comment (pkarashchenko) [inline include/nuttx/clock.h:257]:
```suggestion
      clock_t _tick = (tick); \
```

File: include/nuttx/timers/oneshot.h [hunk 11]
```diff
@@ -192,6 +204,15 @@ struct oneshot_operations_s
...
   CODE int (*current)(struct oneshot_lowerhalf_s *lower,
                       FAR struct timespec *ts);
+  CODE int (*tick_max_delay)(FAR struct oneshot_lowerhalf_s *lower,
+                             FAR clock_t *ticks);
+  CODE int (*tick_start)(FAR struct oneshot_lowerhalf_s *lower,
+                         oneshot_callback_t callback, FAR void *arg,
+                         clock_t ticks);
+  CODE int (*tick_cancel)(struct oneshot_lowerhalf_s *lower,
+                          FAR clock_t *ticks);
+  CODE int (*tick_current)(struct oneshot_lowerhalf_s *lower,
```
Comment (pkarashchenko) [inline include/nuttx/timers/oneshot.h:214]:
```suggestion
  CODE int (*tick_current)(FAR struct oneshot_lowerhalf_s *lower,
```

File: include/nuttx/timers/oneshot.h [hunk 12]
```diff
@@ -192,6 +204,15 @@ struct oneshot_operations_s
                      FAR struct timespec *ts);
   CODE int (*current)(struct oneshot_lowerhalf_s *lower,
                       FAR struct timespec *ts);
+  CODE int (*tick_max_delay)(FAR struct oneshot_lowerhalf_s *lower,
+                             FAR clock_t *ticks);
+  CODE int (*tick_start)(FAR struct oneshot_lowerhalf_s *lower,
+                         oneshot_callback_t callback, FAR void *arg,
+                         clock_t ticks);
+  CODE int (*tick_cancel)(struct oneshot_lowerhalf_s *lower,
```
Comment (pkarashchenko) [inline include/nuttx/timers/oneshot.h:212]:
```suggestion
  CODE int (*tick_cancel)(FAR struct oneshot_lowerhalf_s *lower,
```

File: include/nuttx/timers/oneshot.h [hunk 13]
```diff
@@ -237,6 +258,120 @@ extern "C"
...
+  DEBUGASSERT(lower->ops->tick_cancel);
+
+  ret = lower->ops->tick_cancel(lower, &tick);
+  timespec_from_tick(ts, tick);
+
+  return ret;
+}
+
+static inline
+int oneshot_current(struct oneshot_lowerhalf_s *lower,
```
Comment (pkarashchenko) [inline include/nuttx/timers/oneshot.h:304]:
```suggestion
int oneshot_current(FAR struct oneshot_lowerhalf_s *lower,
```

File: include/nuttx/timers/oneshot.h [hunk 14]
```diff
@@ -237,6 +258,120 @@ extern "C"
...
+  clock_t tick;
+
+  DEBUGASSERT(lower->ops->tick_start);
+
+  tick = timespec_to_tick(ts);
+  return lower->ops->tick_start(lower, callback, arg, tick);
+}
+
+static inline
+int oneshot_cancel(struct oneshot_lowerhalf_s *lower,
```
Comment (pkarashchenko) [inline include/nuttx/timers/oneshot.h:289]:
```suggestion
int oneshot_cancel(FAR struct oneshot_lowerhalf_s *lower,
```

File: include/nuttx/timers/oneshot.h [hunk 15]
```diff
@@ -237,6 +258,120 @@ extern "C"
...
+  struct timespec ts;
+
+  DEBUGASSERT(lower->ops->start);
+
+  timespec_from_tick(&ts, ticks);
+  return lower->ops->start(lower, callback, arg, &ts);
+}
+
+static inline
+int oneshot_tick_cancel(struct oneshot_lowerhalf_s *lower,
```
Comment (pkarashchenko) [inline include/nuttx/timers/oneshot.h:346]:
```suggestion
int oneshot_tick_cancel(FAR struct oneshot_lowerhalf_s *lower,
```

File: include/nuttx/timers/oneshot.h [hunk 16]
```diff
@@ -237,6 +258,120 @@ extern "C"
...
+  DEBUGASSERT(lower->ops->cancel);
+
+  ret = lower->ops->cancel(lower, &ts);
+  *ticks = timespec_to_tick(&ts);
+
+  return ret;
+}
+
+static inline
+int oneshot_tick_current(struct oneshot_lowerhalf_s *lower,
```
Comment (pkarashchenko) [inline include/nuttx/timers/oneshot.h:361]:
```suggestion
int oneshot_tick_current(FAR struct oneshot_lowerhalf_s *lower,
```

File: arch/arm/src/stm32/stm32_tickless.c (hunk 1 above)
Comment (zyfeier) [reply arch/arm/src/stm32/stm32_tickless.c]: Done

File: arch/arm/src/stm32/stm32_tickless.c (hunk 2 above)
Comment (zyfeier) [reply arch/arm/src/stm32/stm32_tickless.c]: Done

File: arch/arm/src/stm32f7/stm32_tickless.c (hunk 3 above)
Comment (zyfeier) [reply arch/arm/src/stm32f7/stm32_tickless.c]: Done

File: arch/arm/src/stm32f7/stm32_tickless.c (hunk 4 above)
Comment (zyfeier) [reply arch/arm/src/stm32f7/stm32_tickless.c]: Done

File: arch/arm/src/stm32h7/stm32_tickless.c (hunk 5 above)
Comment (zyfeier) [reply arch/arm/src/stm32h7/stm32_tickless.c]: Done

**Comment by masayuki2009** [comment]
@zyfeier 
Please rebase this PR to the latest upstream master.

**Pull request merged**